File: python/arrays-and-hashing/sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        return self.sort(nums)

    def sort(self, nums):
        middle = len(nums) // 2

        if len(nums) > 1:
            left = self.sort(nums[middle:])
            right = self.sort(nums[:middle])

            return self.merge(left, right)

        return nums

    def merge(self, left, right):
        merged = []
        l, r = 0, 0
        while l < len(left) and r < len(right):
            if left[l] > right[r]:
                merged.append(right[r])
                r += 1
            else:
                merged.append(left[l])
                l += 1

        while l < len(left):
            merged.append(left[l])
            l += 1

        while r < len(right):
            merged.append(right[r])
            r += 1

        return merged

File: c++/arrays-and-hashing/sort-an-array.cc
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        merge_sort(nums, 0, nums.size() - 1);

        return nums;
    }

private:
    void merge_sort(std::vector<int>& nums, int left, int right) {
        if (right <= left) {
            return;
        }
        
        int middle = left + int((right - left) / 2);

        merge_sort(nums, left, middle);
        merge_sort(nums, middle + 1, right);

        merge(nums, left, middle+1, right);
    }

    void merge(std::vector<int>& nums, int first, int second, int end) {
        std::vector<int> buffer = {};
        int start = first;
        int mid = second - 1;
        while (first <= mid || second <= end) {
            if (first > mid) {
                buffer.push_back(nums[second]);
                second++;
            } else if (second > end) {
                buffer.push_back(nums[first]);
                first++;
            } else if (nums[first] < nums[second]) {
                buffer.push_back(nums[first]);
                first++;
            } else {
                buffer.push_back(nums[second]);
                second++;
            }
        }

        for (int i = start; i <= end; ++i) {
            nums[i] = buffer[i - start];
        }
    }
};
